regparser/api_writer.py

```diff
@@ -123,16 +131,396 @@ def write(self, python_object):
+        else:
+            elem = Element('paragraph', label=root.label_id(), marker=root.label[-1])
+            if root.title:
+                title = SubElement(elem, 'title')
+                title.text = root.title
+            text = self.apply_layers(root)
+            content = fromstring('<content>' + text + '</content>')
+            elem.append(content)
+
+        if len(root.label) > 1 and ('Subpart' not in root.label):
+            # the part is a special case
+            for child in root.children:
+                sub_elem = self.to_xml(child)
+                elem.append(sub_elem)
+        return elem
+
+    def apply_layers(self, node):
+        text = node.text
+        replacement_hashes = {}
+        all_offsets = []
+        all_replacements = []
+        # import pdb; pdb.set_trace()
+        for ident, layer in self.layers.items():
+            # print node.label_id()
+            if node.label_id() in layer:
+                # print 'applying layers'
+                replacements = layer[node.label_id()]
+                if ident == 'terms':
+                    offsets, repls = XMLWriteContent.apply_terms(text, replacements)
+                elif ident == 'paragraph-markers':
+                    offsets, repls = XMLWriteContent.apply_paragraph_markers(text, replacements)
+                elif ident == 'internal-citations':
+                    offsets, repls = XMLWriteContent.apply_internal_citations(text, replacements)
+                elif ident == 'definitions':
+                    offsets, repls = XMLWriteContent.apply_definitions(text, replacements)
+                elif ident == 'external-citations':
+                    offsets, repls = XMLWriteContent.apply_external_citations(text, replacements)
+                else:
+                    offsets = []
+                    repls = []
+
+                all_offsets.extend(offsets)
+                all_replacements.extend(repls)
+
+        if len(all_offsets) > 0 and len(all_replacements) > 0:
+
+            offsets_and_repls = zip(all_offsets, all_replacements)
+            offsets_and_repls = sorted(offsets_and_repls, key=lambda x: x[0][0])
+            all_offsets, all_replacements = zip(*offsets_and_repls)
+
+            text = interpolate_string(text, all_offsets, all_replacements).strip()
+
+        return text
+
 class Client:
     """A Client for writing regulation(s) and meta data."""
 
-    def __init__(self):
-        if settings.API_BASE:
-            self.writer_class = APIWriteContent
-        elif getattr(settings, 'GIT_OUTPUT_DIR', ''):
-            self.writer_class = GitWriteContent
+    def __init__(self, writer_type=None):
+        if writer_type is None:
+            if settings.API_BASE:
+                self.writer_class = APIWriteContent
+            elif getattr(settings, 'GIT_OUTPUT_DIR', ''):
+                self.writer_class = GitWriteContent
+            else:
+                self.writer_class = FSWriteContent
```

Comment:
So, this is the indirection problem that I was talking about. Basically, I think the writer subclasses are really unwieldy and have forced the output code into a spaghetti tangle of indirection. Just check out how they're instantiated and used; it's a nightmare. I had to hack something useful onto it in the fastest way. I do not want to spend a ton of time working out the perfect writer class structure because I don't think it's that relevant and also because in order to do this in a more aesthetically pleasing and comprehensible manner we'd have to rip out that entire framework and refactor it. So, sure, this is not ideal, but it'll do the trick and then hopefully we'll never have to look at it again.